device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionHandler.java

```diff
@@ -333,57 +283,60 @@ boolean handleAuthenticationMessage(AmqpsMessage amqpsMessage)
+        boolean result = false;
+
+        if (this.amqpsAuthenticatorState == AmqpsDeviceAuthenticationState.AUTHENTICATED && hasLink(linkName))
         {
-            for (Map.Entry<MessageType, AmqpsDeviceOperations> entry : amqpsDeviceOperationsMap.entrySet())
+            for (Map.Entry<MessageType, AmqpsLinksHandler> entry : amqpsLinkMap.entrySet())
             {
                 if (entry.getValue().onLinkRemoteOpen(linkName))
                 {
                     // If the link that is being opened is a sender link and the operation is a DeviceTwin operation, we will send a subscribe message on the opened link
                     if (entry.getKey() == MessageType.DEVICE_TWIN && linkName.equals(entry.getValue().getSenderLinkTag()))
                     {
                         // since we have already checked the message type, we can safely cast it
-                        AmqpsDeviceTwin deviceTwinOperations = (AmqpsDeviceTwin)entry.getValue();
+                        AmqpsTwinLinksHandler deviceTwinOperations = (AmqpsTwinLinksHandler)entry.getValue();
                         int deliveryTag = sendMessage(deviceTwinOperations.buildSubscribeToDesiredPropertiesProtonMessage(), entry.getKey(), deviceClientConfig.getDeviceId());
                         this.subscriptionMessageRequestSentCallback.onSubscriptionMessageSent(deliveryTag, SubscriptionMessageRequestSentCallback.SubscriptionType.DESIRED_PROPERTIES_SUBSCRIPTION);
                     }
 
-                    return true;
+                    result = true;
+                }
+            }
+
+            for (Map.Entry<MessageType, AmqpsLinksHandler> entry : amqpsLinkMap.entrySet())
+            {
+                if (!entry.getValue().isOpen())
+                {
+                    return result;
                 }
             }
+
+            sessionOpenedCallback.onDeviceSessionOpened(this.deviceClientConfig.getDeviceId());
         }
 
-        return false;
+        return result;
     }
 
     /**
      * Convert from IoTHub message to Proton using operation 
      * specific converter. 
      *
      * @param message the message to convert.
-     * @throws TransportException if conversion fails.
-     * @return AmqpsConvertToProtonReturnValue the result of the 
-     *         conversion containing the Proton message.
+     * @return The result of the conversion containing the Proton message.
      */
-    AmqpsConvertToProtonReturnValue convertToProton(Message message) throws TransportException
+    MessageImpl convertToProton(Message message)
     {
-        MessageType msgType;
         if (message.getMessageType() == null)
         {
-            msgType = DEVICE_TELEMETRY;
-        }
-        else
-        {
-            msgType = message.getMessageType();
+            message.setMessageType(DEVICE_TELEMETRY);
         }
 
-        if (amqpsDeviceOperationsMap.get(msgType) != null)
+        if (amqpsLinkMap.get(message.getMessageType()) != null)
```

Comment:
seems like before the if statement it should get the value, then compare it to null, and if not null, use the cached value. Right now we have 2 look ups.

Reply:
Sure, this can be simplified